File: MiniDOMSIP/Config/Servers/AsteriskServers.cs
using minidom.CallManagers;

namespace minidom.PBX
{
    public sealed class AsteriskServers
    {
        public static event ConnectedEventHandler Connected;

        public delegate void ConnectedEventHandler(object sender, AsteriskEventArgs e);

        public static event DisconnectedEventHandler Disconnected;

        public delegate void DisconnectedEventHandler(object sender, AsteriskEventArgs e);

        public static event ManagerEventEventHandler ManagerEvent;

        public delegate void ManagerEventEventHandler(object sender, AsteriskEvent e);

        private static CCollection<AsteriskServer> m_Servers = new CCollection<AsteriskServer>();

        private AsteriskServers()
        {
            DMDObject.IncreaseCounter(this);
        }

        static AsteriskServers()
        {
        }

        public static void StartListening(CCollection<AsteriskServer> servers)
        {
            var unchanged = new CCollection<AsteriskServer>();

            // Rimuoviamo i servers presenti in entrambi le configurazioni
            int i = 0;
            while (i < m_Servers.Count)
            {
                var curr = m_Servers[i];
                bool found = false;
                int j = 0;
                while (j < servers.Count)
                {
                    var other = servers[j];
                    if (curr.Equals(other))
                    {
                        unchanged.Add(curr);
                        servers.RemoveAt(j);
                        other.SetManager(curr.GetManager());
                        found = true;
                        break;
                    }
                    else
                    {
                        j += 1;
                    }
                }

                if (found)
                {
                    m_Servers.RemoveAt(i);
                }
                else
                {
                    i += 1;
                }
            }

            // Disconnettiamo le configurazioni non valide
            foreach (AsteriskServer server in m_Servers)
            {
                if (server.IsConnected())
                    server.Disconnect();
                server.Connected -= handleAsteriskConnect;
                server.Disconnected -= handleAsteriskDisconnect;
                server.ManagerEvent -= handleAsteriskEvent;
            }

            // Connettiamo le nuove configurazioni
            foreach (AsteriskServer server in servers)
            {
                server.Connected += handleAsteriskConnect;
                server.Disconnected += handleAsteriskDisconnect;
                server.ManagerEvent += handleAsteriskEvent;
                server.Connect();
            }

            m_Servers = new CCollection<AsteriskServer>();
            m_Servers.AddRange(unchanged);
            m_Servers.AddRange(servers);
        }

        public static void StopListening()
        {
            foreach (AsteriskServer server in m_Servers)
            {
                server.Disconnect();
                server.Connected -= handleAsteriskConnect;
                server.Disconnected -= handleAsteriskDisconnect;
                server.ManagerEvent -= handleAsteriskEvent;
            }

            m_Servers.Clear();
        }

        private static void handleAsteriskEvent(object sender, AsteriskEvent e)
        {
            ManagerEvent?.Invoke(sender, e);
        }

        private static void handleAsteriskDisconnect(object sender, AsteriskEventArgs e)
        {
            Disconnected?.Invoke(sender, e);
        }

        private static void handleAsteriskConnect(object sender, AsteriskEventArgs e)
        {
            Connected?.Invoke(sender, e);
        }

        ~AsteriskServers()
        {
            DMDObject.DecreaseCounter(this);
        }
    }
}

File: MiniDOMSIP/Config/AsteriskServers.vb
Imports System.IO
Imports System.Xml.Serialization
Imports minidom
Imports minidom.Sistema
Imports minidom.CallManagers

Public NotInheritable Class AsteriskServers



    Public Shared Event Connected(ByVal sender As Object, ByVal e As AsteriskEventArgs)
    Public Shared Event Disconnected(ByVal sender As Object, ByVal e As AsteriskEventArgs)
    Public Shared Event ManagerEvent(ByVal sender As Object, ByVal e As AsteriskEvent)

    Private Shared m_Servers As New CCollection(Of AsteriskServer)

    Private Sub New()
        DMDObject.IncreaseCounter(Me)
    End Sub

    Shared Sub New()
    End Sub


    Public Shared Sub StartListening(ByVal servers As CCollection(Of AsteriskServer))
        Dim unchanged As New CCollection(Of AsteriskServer)

        'Rimuoviamo i servers presenti in entrambi le configurazioni
        Dim i As Integer = 0
        While (i < m_Servers.Count)
            Dim curr As AsteriskServer = m_Servers(i)
            Dim found As Boolean = False
            Dim j As Integer = 0
            While (j < servers.Count)
                Dim other As AsteriskServer = servers(j)
                If (curr.Equals(other)) Then
                    unchanged.Add(curr)
                    servers.RemoveAt(j)
                    other.SetManager(curr.GetManager)
                    found = True
                    Exit While
                Else
                    j += 1
                End If
            End While
            If (found) Then
                m_Servers.RemoveAt(i)
            Else
                i += 1
            End If
        End While

        'Disconnettiamo le configurazioni non valide
        For Each server As AsteriskServer In m_Servers
            If server.IsConnected Then server.Disconnect()
            RemoveHandler server.Connected, AddressOf handleAsteriskConnect
            RemoveHandler server.Disconnected, AddressOf handleAsteriskDisconnect
            RemoveHandler server.ManagerEvent, AddressOf handleAsteriskEvent
        Next

        'Connettiamo le nuove configurazioni
        For Each server As AsteriskServer In servers
            AddHandler server.Connected, AddressOf handleAsteriskConnect
            AddHandler server.Disconnected, AddressOf handleAsteriskDisconnect
            AddHandler server.ManagerEvent, AddressOf handleAsteriskEvent
            server.Connect()
        Next

        m_Servers = New CCollection(Of AsteriskServer)
        m_Servers.AddRange(unchanged)
        m_Servers.AddRange(servers)
    End Sub

    Public Shared Sub StopListening()
        For Each server As AsteriskServer In m_Servers
            server.Disconnect()
            RemoveHandler server.Connected, AddressOf handleAsteriskConnect
            RemoveHandler server.Disconnected, AddressOf handleAsteriskDisconnect
            RemoveHandler server.ManagerEvent, AddressOf handleAsteriskEvent
        Next
        m_Servers.Clear()
    End Sub


    Private Shared Sub handleAsteriskEvent(sender As Object, e As AsteriskEvent)
        RaiseEvent ManagerEvent(sender, e)
    End Sub

    Private Shared Sub handleAsteriskDisconnect(sender As Object, e As AsteriskEventArgs)
        RaiseEvent Disconnected(sender, e)
    End Sub

    Private Shared Sub handleAsteriskConnect(sender As Object, e As AsteriskEventArgs)
        RaiseEvent Connected(sender, e)
    End Sub

    Protected Overrides Sub Finalize()
        MyBase.Finalize()
        DMDObject.DecreaseCounter(Me)
    End Sub
End Class
